### ISSUE TITLE
Feature request: Add event handler in LogManager that is invoked in case of internal NLog/Extension exceptions

### ISSUE THREAD

**Issue by katdti** [NLog/NLog#2710]
Hi

If a logging target fails due to socket/http connection errors - it should be possible to detect this from  apps using NLog - this would allow the app to e.g. inform the user via the GUI,
The app should be able to attach to an event handler and get a callback first time problem is detected and a callback when the connection problem is solved and first successfully log to server was performed.

Optimal would be:
- if the internal logging stop logging the same error until the error disappears.
- any failed writes are cached until connection is back online

**Additional Information**:

NLog version: 4.5.3
Platform: Net 4.5

Sample configuration from PC that genrated an exception due to https sequity rules on local PC.
Exception comes at almost every logging.
```xml
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
        throwExceptions="false" internalLogLevel="Warn" internalLogToConsole="false" autoReload="false"
        internalLogFile="C:/Temp/communications.log"
    >
    <extensions>
         <add assembly="NLog.Targets.Seq"/>
    </extensions>
<targets async="true">
      <target name="seq" xsi:type="Seq" serverUrl="https://localhost:5341" apiKey="">
      </target> 
</targets>
    <rules>
      <logger name="*" minlevel="Trace" writeTo="seq" />
    </rules>
</nlog>
```

The exception generated:
```
Error has been raised. Exception: System.Net.WebException: The underlying connection was closed: An unexpected error occurred on a send. ---> System.IO.IOException: The handshake failed due to an unexpected packet format.
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ForceAuthentication(Boolean receiveFirst, Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessAuthentication(LazyAsyncResult lazyResult)
   at System.Net.TlsStream.CallProcessAuthentication(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Net.TlsStream.ProcessAuthentication(LazyAsyncResult result)
   at System.Net.TlsStream.Write(Byte[] buffer, Int32 offset, Int32 size)
   at System.Net.PooledStream.Write(Byte[] buffer, Int32 offset, Int32 size)
   at System.Net.ConnectStream.WriteHeaders(Boolean async)
   --- End of inner exception stack trace ---
   at System.Net.HttpWebRequest.GetResponse()
   at NLog.Targets.Seq.SeqTarget.PostBatch(IEnumerable`1 events)
   at NLog.Targets.Seq.SeqTarget.Write(IList`1 logEvents)
   at NLog.Targets.Target.Write(AsyncLogEventInfo[] logEvents)
   at NLog.Targets.Target.WriteAsyncThreadSafe(IList`1 logEvents)
```

**Comment by 304NotModified** [comment]
Hi,

NLog is route based and not event based.

But you could use the methodcall target: https://github.com/NLog/NLog/wiki/MethodCall-target

**Comment by katdti** [comment]
Route based principle is fine. 

As I understand MethodCall target - method is called every time i make a log.Warning("Hi").
But the request is not to get events for each log - but an event when NLogs internal log writes -  in this case an internal NLog exception. I do not see how MethodCall can capture internal NLog logs.

To day I can set the throwException flag true in the configuration file. This will allow me to catch (some) NLog errors and inform the user via GUI. But it would require a try/catch around every log line in the code - this does not seem as a good solution. 

LogManager has event handlers we can use to detect configuration file changed. The request is to get a handler for "internal NLog/Extension errors" so we can inform user that logging is not working - and perhaps disable the target that is not working.

**Comment by 304NotModified** [comment]
yes your right, I missed the point that it was the internal log.

**Comment by 304NotModified** [comment]
As far as I remember there is already an issue over here with the same request (can't find it now) and the progress was stopped in the design phase.

**Comment by 304NotModified** [comment]
Feel free to send a PR, it would be a nice feature for NLog 4.6

**Comment by snakefoot** [comment]
One could consider something like fallback-wrapper, where one could access the original exception that caused the fallback.

**Comment by katdti** [comment]
I do not know the internals of NLog - but I like the keyword "fallback" in the fallback-wrapper suggestion.
You might not get an exception incase of IO bottleneck. E.g.: I assume an async-wrapper would consume memory or drop messages without raising an exception. And with the option for informing the user there is problems. Maybe it should be a "i am in trouble" mechanism.

**Comment by snakefoot** [comment]
Good point. We should provide exception object when discarding messages in the async wrapper.


Sent from my Samsung device


-------- Original message --------

**Comment by katdti** [comment]
If async wrapper should provide exception object when discarding messages - when I think RetryingWrapper and BufferingWrapper also supply exception objects.

**Comment by snakefoot** [comment]
RetryingWrapper is already providing exception-object on final failure, when no more retries left. But yes BufferingWrapper should also have a reusuable "discard"-exception (Thought i can only discard logevents with special configuration).

What you have to think about is the interface for how to intercept these exceptions. Should a new target-interface be invented that allows one to receive failed logevents (together with org-target and org-exception) ? Should it be a global event-handler on the fallbackwrapper?

If going down the lane with a new target-interface, that is only recognized by the fallbackwrapper, which will then have special logic to call a specialized target-method that takes org-logevent + org-exception + org-target as input. By making use of the little known feature called target-ref, then all fallbackwrappers could be doing fallback to the same global instance of the new target-interface.

**Comment by katdti** [comment]
>>What you have to think about is the interface for how to intercept these exceptions
I have made no thoughts about this yet - it do need some thinking i guess. 
(Internal prioritizing cause me to pause my thinking for the near feature.)

**Comment by snakefoot** [comment]
Alternative then you could attach a custom TextWriter to the InternalLogger, and configure the InternalLogger-LogLevel to Warn/Error. See `LogWriter` here https://github.com/NLog/NLog/wiki/Internal-Logging

Then forward the output from the TextWriter to some place where the user can see it. Then add whatever filtering logic that you might want.

**Comment by katdti** [comment]
Yes, attaching a custom TextWriter to the InternalLogger would make a basic option for warning user via UI.
But feltering requires text pattern matching because textwrite only gets a sting. TextWriter does not get structured information about log level etc.
But it could be a starting point to use a TextWriter.
If InternalLogger had an event with event args like:
   - Message, level, time etc
   - Target
   - Exception Object maybe

it would be posible to generaete better diagnostic messages to the user.

**Comment by snakefoot** [comment]
@katdti It would be a good idea to introduce a InternalLogger-delegate-destination.

The delegate could be like this:

-  `void InternalLogEvent(Type senderType, string senderName, LogLevel, Exception, string logMessage)`

Then create some internal helper methods, that is called like this:

`InternalLogger.Warn(Target sender, Exception exception, string message);`

The internal helper methods will only lookup the "senderType" and "senderName" if the LogLevel is enabled. Maybe implemented together with #2508

**Comment by chucklu** [comment]
Any update on this topic, log4net has a property to handle internal error of the component, [TextWriterAppender.ErrorHandler Property](https://logging.apache.org/log4net/log4net-1.2.13/release/sdk/log4net.Appender.AppenderSkeleton.ErrorHandler.html).
```C#
public override IErrorHandler ErrorHandler { get; set; }
```

**Comment by katdti** [comment]
Additional thoughts. 
Another way of handling exception is to make a new RetryingWrapper that supports a delegate returning e.g. "Retry in x seconds" or "Do not retry" allowing the application to make inteligent retry logic. The application could call a "RetryNow()" method  when it thinks the problem has been solved and it makes sence for the target to retry.

**Comment by snakefoot** [comment]
@chucklu There is `InternalLogger.LogMessageReceived`  See also #3796 

With NLog 5.0 then hopefully #2508 will be implemented, and then InternalLogger will provide Target-context-type and exception-info.

**Comment by chucklu** [comment]
@snakefoot  I am trying to get the file permission problem with NLog, I supposed to get the exception from NLog.
Also tried https://github.com/NLog/NLog/wiki/Internal-Logging, and no log when write log failed with permission problem.
I can handle the permission error from [TextWriterAppender.ErrorHandler](https://logging.apache.org/log4net/release/sdk/html/P_log4net_Appender_TextWriterAppender_ErrorHandler.htm) Property of log4net.
@304NotModified  Is there a way to get the file access denied exception which is caused by lack of permission? I would like to record the permission exception in to event log of windows.

**Comment by chucklu** [comment]
Tried to enable the internal log with configuration file, and it works ,I can find the permission problem in log file.
However, there are two many other log info except the UnauthorizedAccessException.
What I supposed to get, is the only exception log.

**Comment by snakefoot** [comment]
@chucklu Any reason why you cannot use `InternalLogger.LogMessageReceived` ?:

```c#
                NLog.Common.InternalLogger.LogMessageReceived += (sender, e) =>
                {
                    if (e.Exception != null && e.Level > LogLevel.Warn)
                    {
                        if (e.Exception is UnauthorizedAccessException)
                        {
                            // Do some stuff
                        }
                    }
                };
```

**Comment by chucklu** [comment]
@snakefoot Event written to the internal log. Please note that the event is not triggered when then event hasn't the minimal log level set by NLog.Common.InternalLogger.LogLevel.
According the above comments, I tried to set 
```C#
     InternalLogger.LogLevel = NLog.LogLevel.Trace;
```
The problem is, **the LogMessageReceived event is not triggered**, although there should be an unauthorized exception.

**Comment by chucklu** [comment]
@snakefoot  It's my mistake, I am doing the wrong test. The code has block to register the event, and I forget to invoke the code block. That's why the event is not triggerred.

**Comment by snakefoot** [comment]
@chucklu I'm not sitting next to you, so you have to explain what you have tried, before saying it is not working. Does your `NLog.config` by chance update `internalLogLevel="..."` ? If you have a case where you get output in the internal-logfile/console but without `LogMessageReceived` being triggered, then it sounds like a bug.

**Comment by chucklu** [comment]
@snakefoot  internalLogLevel="Off|Trace|Debug|Info|Warn|Error|Fatal" – determines internal log level. The higher the level, the less verbose the internal log output.
I did not quite understanding the above explanation.  For example, when I invoke Logger.Debug and encounter unauthorized exception. What's the log level of internal log I supposed to set,  to trigger the InternalLogger.LogMessageReceived event?

**Comment by chucklu** [comment]
Just do a test, set InternalLogger.LogLevel to Error, and invoke **Logger.Info** method to record the log.
And I can still get the unauthorized exception through `InternalLogger.LogMessageReceived` event. **So the log level of InternalLogger has nothing to do with the invoked method of Logger**, maybe you can make this clear in documentation, I was confused about this before I do the test.  

Then the question would be, which level I supposed to set if I only want to receive the exception of internal error of NLog?

**Comment by snakefoot** [comment]
`InternalLogger` is never activated if called NLog [Logger.Debug](https://nlog-project.org/documentation/v4.7.0/html/M_NLog_Logger_Debug_2.htm)-method.

`InternalLogger` is a diagnostic tool for NLog Targets (Like the FileTarget) to report events. `InternalLogger` doesn't inspect the exceptions logged by the application itself using NLog Logger.

But if you did mean `InternalLogger.Debug(ex, "My Exception)` then it will only be reported to output when using `InternalLogger.LogLevel = NLog.LogLevel.Trace`

**Comment by chucklu** [comment]
> If a Logger.Debug call reaches a NLog Target that triggers an error (Ex. UnauthorizedAccessException), then the event is usually reported to the InternalLogger using another severity-level (Ex Error-LogLevel).  

Good to know this. You might add this to the documentation https://github.com/NLog/NLog/wiki/Internal-Logging .

**Comment by chucklu** [comment]
@snakefoot  I did another test, invoke [Logger.Info](https://nlog-project.org/documentation/v4.7.0/html/Overload_NLog_Logger_Info.htm) method to record log. And set the log level of InternalLogger to Info. And there is no unauthorized exception, I already grant the file access.
Then I find the LogMessageReceived event is not triggered, so it means So _the log level of InternalLogger has nothing to do with the invoked method of Logger_, **could you confirm this**?
I also find the archive log file operation will triggered the LogMessageReceived event, it would be confused. What's the exact purpose of the LogMessageReceived? Who will trigger this event?  and when?

What I try to find is just a simple feature like [TextWriterAppender.ErrorHandler](https://logging.apache.org/log4net/release/sdk/html/P_log4net_Appender_TextWriterAppender_ErrorHandler.htm
) in log4net which is designed only expose the internal exception.
Single-responsibility principle is very important, it would be better if you can provide a new event or handler only used to provide the inner exception of NLog.

**Comment by snakefoot** [comment]
> Single-responsibility principle is very important, it would be better if you can provide a new event or handler only used to provide the inner exception of NLog.

Also a big fan of single-responsibility. And think that `InternalLogger.LogMessageReceived` is a nice implementation of that. Simple with no surprises with funny filtering rules. It is just a raw firehose of events from the NLog engine. Giving the same expected output as `InternalLogger.LogToFile` and `InternalLogger.LogToConsole`.

If suddenly want to have custom filtering like:
- Only want exceptions
- Only want events from a specific NLog-Target-Type
- Only want events that contain a specific word.

Then it sounds like filtering-extensions to the raw firehose. Implementing all kind of exciting filtering-logic on top of the raw firehose might be more confusing for the NLog-users (Suddenly you have 4 event-handles all of them reporting events from NLog engine, which one to choose?). Like to have a simple door that all users can use, instead of having a custom door for each user.

The `InternalLogger.LogMessageReceived` is rather new, and is great to get feedback on this feature. But before deciding that it is completely useless implementation, then I think we should stick with it for a while and collect suggestions for improvements. Then if collection gets big enough, then decide if a version 2 of the interface should be implemented.

**Comment by chucklu** [comment]
@snakefoot You can check SelflLog implemented in Serilog which is used to [debugging and diagnosing](https://github.com/serilog/serilog/wiki/Debugging-and-Diagnostics)
https://github.com/serilog/serilog/blob/dev/src/Serilog/Debugging/SelfLog.cs#L96
It only used to report exception and error, you can check the referenced of the WriteLine method of SelfLog to verify it.
At least the report of exception and error should be an independent responsibility.

**Comment by snakefoot** [comment]
There is no guarantee that Serilog SelfLog will  only report errors. Also SelfLog is just like NLog InternalLogger. LogWriter.


Sent from my Sony Xperia


---- Chuck Lu wrote ----


@snakefoot<https://github.com/snakefoot> You can check SelflLog implemented in Serilog which is used to debugging and diagnosing<https://github.com/serilog/serilog/wiki/Debugging-and-Diagnostics>
https://github.com/serilog/serilog/blob/dev/src/Serilog/Debugging/SelfLog.cs#L96
It only used to report exception and error, you can check the referenced of the WriteLine method of SelfLog to verify it.
At least the report of exception and error should be an independent responsibility.

**Comment by chucklu** [comment]
@snakefoot I have cloned the Serilog and check the source code. Although SelfLog is public class and WriteLine is also a public method, the inner invoke of WriteLine  method are about exception and errors. Maybe this method could be internal.
The public usage is only occur in unit test https://github.com/serilog/serilog/blob/dev/test/Serilog.Tests/Debugging/SelfLogTests.cs

**Comment by snakefoot** [comment]
Think you are discussing things outside the scope of this NLog issue. You probably right that Serilog-Core only reports errors (at the moment) but the same method is used by all Serilog-sinks (in other repositories). Again you have no guarantee that it will only be invoked on errors only. Anyway start new issue if you want to discuss Serilog SelfLog.


Sent from my Sony Xperia


---- Chuck Lu wrote ----


@snakefoot<https://github.com/snakefoot> I have cloned the Serilog and check the source code. Although SelfLog is public class and WriteLine is also a public method, the inner invoke of WriteLine method are about exception and errors. Maybe this method could be internal.
The public usage is only occur in unit test https://github.com/serilog/serilog/blob/dev/test/Serilog.Tests/Debugging/SelfLogTests.cs

**Comment by snakefoot** [comment]
Closing this a resolved by #3796 as you can get a callback on errors. 

More improvements to providing more target-context-details will come with NLog 5.0,
